pennylane/gradients/param_shift_hessian.py

```diff
@@ -0,0 +1,256 @@
+    for i in range(h_dim):
+        for j in range(h_dim):
+            # optimization: only generate tapes for upper triangular matrix (j >= i)
+            # optimization: skip partial derivates that are zero
+            if j < i or diff_methods[i] == "0" or diff_methods[j] == "0":
+                gradient_coeffs.append([])
+                shapes.append(((i, j), 0))
+                continue
+
+            recipe = diag_recipe if i == j else off_diag_recipe
+            coeffs, _, shifts = _process_gradient_recipe(recipe)
+
+            # optimization: only compute the unshifted tape once
+            if all(np.array(shifts[0]) == 0):
+                if not unshifted_coeffs and f0 is None:
+                    gradient_tapes.insert(0, tape)
+
+                unshifted_coeffs[(i, j)] = coeffs[0]
+                coeffs, shifts = coeffs[1:], shifts[1:]
+
+            # generate the gradient tapes
+            gradient_coeffs.append(coeffs)
+            g_tapes = generate_multishifted_tapes(tape, (i, j), shifts)
+
+            gradient_tapes.extend(g_tapes)
+            shapes.append(((i, j), len(g_tapes)))
+
+    def processing_fn(results):
+        # The first results dimension is the number of terms/tapes in the parameter-shift
+        # rule, the remaining ones are the QNode output dimensions.
+        out_dim = qml.math.shape(results)[1:]
+        # The desired shape of the Hessian is:
+        #       (QNode output dimensions, # trainable gate args, # trainable gate args),
+        # but first we accumulate all elements into a list, since no array assingment is possible.
+        hessian = []
+        # Keep track of tape results already consumed.
+        start = 1 if unshifted_coeffs and f0 is None else 0
+        # Results of the unshifted tape.
+        r0 = results[0] if start == 1 else f0
+
+        for k, ((i, j), s) in enumerate(shapes):
+            res = results[start : start + s]
+            start = start + s
+
+            # Compute the elements of the Hessian as the linear combination of
+            # results and coefficients, barring optimization cases.
+            if j < i:
+                g = hessian[j * h_dim + i]
+            elif s == 0:
+                g = qml.math.zeros(out_dim)
+            else:
+                res = qml.math.stack(res)
+                g = qml.math.tensordot(
+                    res, qml.math.convert_like(gradient_coeffs[k], res), [[0], [0]]
+                )
+                if (i, j) in unshifted_coeffs:
+                    g += unshifted_coeffs[(i, j)] * r0
+
+            hessian.append(g)
+
+        # Reshape the Hessian to have the dimensions of the QNode output on the outside, that is:
+        #         (h_dim, h_dim, out_dim) -> (out_dim, h_dim, h_dim)
+        hessian = qml.math.reshape(qml.math.stack(hessian), (h_dim, h_dim) + out_dim)
+        dim_indices = list(range(len(out_dim) + 2))
+        hessian = qml.math.transpose(hessian, axes=dim_indices[2:] + [0, 1])
```

Comment:
```suggestion
        dim_indices = list(range(2, len(out_dim) + 2))
        hessian = qml.math.transpose(hessian, axes=dim_indices + [0, 1])
```

Reply:
This does make sense as an optimization, but the change doesn't match what the variable `dim_indices` represents.
But I can change it to:
```suggestion
        reordered_indices = list(range(2, len(out_dim) + 2)) + [0, 1]
        hessian = qml.math.transpose(hessian, axes=reordered_indices)
```

Reply:
Ah, yeah of course! :) Not an important change of course! :)